Make paginated calls to the Github GraphQL API

digitalronin opened the pull request:
This fixes a problem for teams who have more than 100 github repositories (e.g. the webops team, who have 110). 

The PR bot was only fetching open PRs for the first 100 repositories. This commit fixes that by using the pageInfo returned by the API to figure out when there are more repositories' PRs to fetch, and looping until all data has been retrieved.

This PR also splits the bot source code, separating the web server, PR fetcher, and output formatter into separate files. It adds a Dockerfile and instructions for running the bot locally, for development. This should also make it easier to move the bot into the cloud platform kubernetes cluster, when time permits.

I have tested this PR by briefly telling Heroku to deploy the bot from this branch, and then invoking it in a new Slack channel, and it works as expected.

Commit:
Switch to ruby 2.6.2 (from 2.3.0) and update gems

Commit:
Reorganise code and add paged github API calls

* Move the PR fetcher to a separate class
* Move the output formatter to a separate class
* Repeat the github API call until no more data

The cloud platform team has >100 github repos, but
the API call was only fetching open PRs for the
first 100 repositories.

This change uses the returned 'pageInfo' data to
repeat the API call until there is no more data to
retrieve.

Commit:
Correct the list of env. vars. in the README

Commit:
Remove trailing whitespace

Commit:
Dockerise the PR bot application, for development

Commit:
Git ignore .env file

SteveMarshall commented on lib/pr/fetcher.rb:
```diff
@@ -0,0 +1,145 @@
+        end
+
+        {
+          "Authorization" => "Bearer #{ENV['GH_TOKEN']}"
```
You're setting a `token` variable a few lines before this; should you be using that here?

digitalronin replied:
Good point. I missed that one.

SteveMarshall commented on makefile:
```diff
@@ -0,0 +1,37 @@
+.built-docker-image: Dockerfile Gemfile config.ru web.rb lib/pr.rb $(wildcard lib/pr/*.rb)
+	docker build -t $(IMAGE) .
+	docker tag $(IMAGE) $(TAGGED)
+	touch .built-docker-image
```
It'd be nice to have a corresponding `make clean` that removes the built docker image and the file you touch here.

digitalronin replied:
Good idea.

SteveMarshall reviewed (approved):
This is really good, thanks! A couple of very minor suggestions, but I'm happy for you to fix those later if you'd rather.

I'd also have liked to have seen 947a814a broken up into a no-op “extract to classes” followed by a commit (or two?) adding the new functionality. As it is, it's a bit hard to see which bits are new and which bits are just moving around of existing, if you see what I mean? It's fine here, but worth bearing in mind for future.

Commit:
Add makefile targets

* build

This is here to be the first makefile target.
Without it, the .built-docker-image target is not
counted as the first target, so a bare 'make' will
try to push the docker image to docker hub.

* clean

Remove the docker image and the 
.built-docker-image file

Commit:
Use 'token' variable, not ENV['GH_TOKEN']

This variable was declared, but was not being used
after the initial test for existence

digitalronin commented:
> I'd also have liked to have seen 947a814 broken up

Yes, that's exactly how I should have done it. I did a bunch of stuff while figuring out how it all worked, mixed in with pulling the structure apart (also as a part of figuring it out). 

I did think about disentangling this, to make the history clear, but I was already getting concerned about the amount of time I'd spent on this, so I decided to just push it as was.
